### PR TITLE
Omg 290 investigate state corruption cases in load tests and disaster recovery scenarios

### PR DESCRIPTION
**Pull request by pthomalla**

### CODE REVIEW

File: apps/omg_watcher/lib/block_getter.ex [hunk 1]
```diff
@@ -123,29 +114,19 @@ defmodule OMG.Watcher.BlockGetter do
...
 
   def handle_info({_ref, {:downloaded_block, response}}, state) do
     # 1/ process the block that arrived and consume
     {continue, new_state, events} = Core.handle_downloaded_block(state, response)
-
     EventerAPI.emit_events(events)
 
     with :ok <- continue do
-      {:noreply, new_state}
+      _ = Enum.map(events, fn %Event.InvalidBlock{number: number} -> download_block(number) end)
```
Comment (pdobacz) [inline apps/omg_watcher/lib/block_getter.ex]: why this?

File: apps/omg_watcher/lib/block_getter/core.ex [hunk 2]
```diff
@@ -250,7 +237,7 @@ defmodule OMG.Watcher.BlockGetter.Core do
       )
 
     blocks_numbers =
-      (potential_block_withholding_numbers ++ potential_next_block_numbers)
+      potential_next_block_numbers
```
Comment (pdobacz) [inline apps/omg_watcher/lib/block_getter/core.ex]: this is unexpected - we want the block numbers marked as potential withholding to be re-downloaded

File: apps/omg_watcher/lib/block_getter/core.ex [hunk 3]
```diff
@@ -283,35 +270,29 @@ defmodule OMG.Watcher.BlockGetter.Core do
         ) ::
           {:ok | {:needs_stopping, block_error()}, %__MODULE__{},
            [] | list(Event.InvalidBlock.t()) | list(Event.BlockWithholding.t())}
-          | {:error, :duplicate | :unexpected_blok}
+          | {:error, :duplicate | :unexpected_block}
   def handle_downloaded_block(
-        %__MODULE__{number_of_blocks_being_downloaded: number_of_blocks_being_downloaded} = state,
-        response
+        %__MODULE__{
```
Comment (pdobacz) [inline apps/omg_watcher/lib/block_getter/core.ex]: it seems like you're reverting a previous bug fix. If a withholding is detected you should decrease the number of blocks being downloaded (i.e. a slot is free now).

File: apps/omg_watcher/test/block_getter/core_test.exs [hunk 4]
```diff
@@ -26,70 +26,74 @@ defmodule OMG.Watcher.BlockGetter.CoreTest do
...
-    assert block_numbers == [1_000, 2_000, 3_000, 4_000]
 
-    state_after_proces_down =
-      state_after_chunk
-      |> handle_downloaded_block(%Block{number: 4_000})
-      |> handle_downloaded_block(%Block{number: 2_000})
-
-    assert {_, [5_000, 6_000]} = Core.get_numbers_of_blocks_to_download(state_after_proces_down, 20_000)
+    start_block_number
+    |> Core.init(_interval = 1_000, _synced_height = 1, maximum_number_of_pending_blocks: 4)
```
Comment (pdobacz) [inline apps/omg_watcher/test/block_getter/core_test.exs]: I think this `_interval = 1_000` isn't standard syntax (is it?) and is quite surprising, pls revert. Also note that it is now allowed to start `|>` with a function call.

Multiple times in the PR

File: apps/omg_watcher/test/block_getter/core_test.exs [hunk 5]
```diff
@@ -224,168 +220,128 @@ defmodule OMG.Watcher.BlockGetter.CoreTest do
...
-
-    assert {:ok, state, []} = Core.handle_downloaded_block(state, {:ok, %Block{number: 3_000}})
-
-    assert {_, [5000, 6000, 7000, 8000]} = Core.get_numbers_of_blocks_to_download(state, 20_000)
+      |> Core.handle_downloaded_block(Core.validate_download_response({:error, :error_reason}, <<>>, 3_000, 0))
+      |> assert_check(:ok, [])
+
+    state
+    |> Core.get_numbers_of_blocks_to_download(20_000)
+    |> assert_check([5_000, 6_000])
```
Comment (pdobacz) [inline apps/omg_watcher/test/block_getter/core_test.exs]: thanks to the huge refactor mixed with behavior changes, this is really hard to spot and makes review a nightmare. Please don't do that. Are there any more non-refactor changes to these tests? 

also this block `3_000` missing seems like exactly the bug that was fixed - a failed block download should be retried on first attempt. What is going on here?

File: apps/omg_watcher/test/block_getter/core_test.exs [hunk 6]
```diff
@@ -224,168 +220,128 @@ defmodule OMG.Watcher.BlockGetter.CoreTest do
...
+    block_height
+    |> Core.init(_interval = 1_000, _synced_height = 1, maximum_number_of_pending_blocks: 3)
+    |> Core.get_numbers_of_blocks_to_download(20_000)
+    |> assert_check([1_000, 2_000, 3_000])
+    |> Core.handle_downloaded_block(Core.validate_download_response({:error, :error_reason}, <<>>, 1_000, 0))
+    |> assert_check(:ok, [])
+    |> Core.handle_downloaded_block(Core.validate_download_response({:error, :error_reason}, <<>>, 2_000, 0))
+    |> assert_check(:ok, [])
+    |> Core.get_numbers_of_blocks_to_download(20_000)
+    |> assert_check([])
```
Comment (pdobacz) [inline apps/omg_watcher/test/block_getter/core_test.exs]: same here, blocks 1, 2, 3 should be retried, why is this being removed?

File: apps/omg_watcher/test/block_getter/core_test.exs [hunk 7]
```diff
@@ -224,168 +220,128 @@ defmodule OMG.Watcher.BlockGetter.CoreTest do
...
+    |> Core.init(
+      _interval = 1_000,
+      _synced_height = 1,
+      maximum_number_of_pending_blocks: 2,
+      maximum_block_withholding_time_ms: 10_000
+    )
+    |> Core.get_numbers_of_blocks_to_download(20_000)
+    |> assert_check([1_000, 2_000])
+    |> Core.handle_downloaded_block(potential_withholding_1_000)
+    |> assert_check(:ok, [])
```
Comment (pdobacz) [inline apps/omg_watcher/test/block_getter/core_test.exs]: again, why aren't the withheld blocks retried???

File: apps/omg_watcher/test/block_getter/core_test.exs [hunk 8]
```diff
@@ -202,80 +204,76 @@ defmodule OMG.Watcher.BlockGetter.CoreTest do
     assert {:ok, %{number: 2 = _overridden_number}} = Core.validate_download_response({:ok, block}, hash, 2, 0, 0)
   end
 
-  test "got_block function called once with PotentialWithholding doesn't return BlockWithholding event" do
-    {state, [1_000, 2_000]} = init_state() |> Core.get_numbers_of_blocks_to_download(3_000)
-
-    potential_withholding = Core.validate_download_response({:error, :error_reason}, <<>>, 2_000, 0, 0)
-
-    assert {:ok, _, []} = Core.handle_downloaded_block(state, potential_withholding)
+  test "got_block function called once with PotentialWithholdingReport doesn't return BlockWithholding event" do
```
Comment (pgebal) [inline apps/omg_watcher/test/block_getter/core_test.exs]: There is no got_block function.

File: apps/omg_watcher/test/block_getter/core_test.exs [hunk 9]
```diff
@@ -202,80 +204,76 @@ defmodule OMG.Watcher.BlockGetter.CoreTest do
...
+    init_state()
+    |> Core.get_numbers_of_blocks_to_download(3_000)
+    |> assert_check([1_000, 2_000])
+    |> Core.handle_downloaded_block(Core.validate_download_response({:error, :error_reason}, <<>>, 2_000, 0, 0))
```
Comment (pgebal) [inline apps/omg_watcher/test/block_getter/core_test.exs]: I would explicitly check if we have PotentialWithholdingReport return here.

File: apps/omg_watcher/test/block_getter/core_test.exs [hunk 10]
```diff
@@ -202,80 +204,76 @@ defmodule OMG.Watcher.BlockGetter.CoreTest do
...
-
-    assert {:ok, state, []} = Core.handle_downloaded_block(state, {:ok, %Block{number: 3_000}})
+    init_state(opts: [maximum_number_of_pending_blocks: 4, maximum_block_withholding_time_ms: 0])
+    |> Core.get_numbers_of_blocks_to_download(20_000)
+    |> assert_check([1_000, 2_000, 3_000, 4_000])
+    |> handle_downloaded_block(%Block{number: 1_000})
+    |> handle_downloaded_block(%Block{number: 2_000})
+    |> Core.handle_downloaded_block(Core.validate_download_response({:error, :error_reason}, <<>>, 3_000, 0, 0))
+    |> assert_check(:ok, [])
+    |> Core.get_numbers_of_blocks_to_download(5_000)
```
Comment (pgebal) [inline apps/omg_watcher/test/block_getter/core_test.exs]: What happened with this?
```
assert {_, [3000, 5000, 6000]} = Core.get_numbers_of_blocks_to_download(state, 20_000)
```

File: apps/omg_watcher/test/block_getter/core_test.exs [hunk 11]
```diff
@@ -202,80 +204,76 @@ defmodule OMG.Watcher.BlockGetter.CoreTest do
...
+    |> assert_check([3_000])
+    |> handle_downloaded_block(%Block{number: 3_000})
+    |> Core.get_numbers_of_blocks_to_download(20_000)
+    |> assert_check([5_000, 6_000, 7_000])
+  end
 
-    assert {_, [5000, 6000, 7000, 8000]} = Core.get_numbers_of_blocks_to_download(state, 20_000)
+  test "get_numbers_of_blocks_to_download don't send block that are download" do
```
Comment (pgebal) [inline apps/omg_watcher/test/block_getter/core_test.exs]: get_number_of_blocks_to_download does not return blocks that are being downloaded

File: apps/omg_watcher/test/block_getter/core_test.exs [hunk 12]
```diff
@@ -202,80 +204,76 @@ defmodule OMG.Watcher.BlockGetter.CoreTest do
...
+    |> assert_check([1_000, 2_000, 3_000, 4_000])
+    |> handle_downloaded_block(%Block{number: 1_000})
+    |> handle_downloaded_block(%Block{number: 2_000})
+    |> Core.handle_downloaded_block(Core.validate_download_response({:error, :error_reason}, <<>>, 3_000, 0, 0))
+    |> assert_check(:ok, [])
+    |> Core.get_numbers_of_blocks_to_download(20_000)
+    |> assert_check([3_000, 5_000, 6_000])
+    |> handle_downloaded_block(%Block{number: 5_000})
+    |> Core.get_numbers_of_blocks_to_download(20_000)
+    |> assert_check([7_000])
```
Comment (pgebal) [inline apps/omg_watcher/test/block_getter/core_test.exs]: Please test what happens when one potentialy withheld block download fails. Is it returned in the next call to Core.get_numbers_of_blocks_to_download?

**Pull request merged**